src/Entity/SmsDeliveryReport.php

```diff
@@ -258,11 +235,29 @@ protected function getRevisionAtStatus($status) {
       ->range(0, 1)
       ->execute();
     if ($revision_ids) {
-      return $storage->load(array_keys($revision_ids)[0]);
+      return $storage->loadRevision(key($revision_ids));
     }
-    else {
-      return NULL;
+    return NULL;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function preSave(EntityStorageInterface $storage) {
+    // SMS delivery report cannot be saved without a parent SMS message.
+    if (!$this->getSmsMessage()) {
+      throw new \LogicException('No parent SMS message specified for SMS delivery report');
```

Comment:
This should even be an assertion, going by https://www.drupal.org/docs/8/api/runtime-assertions